File: modules/tiktok-business/android/src/main/java/expo/modules/tiktokbusiness/TiktokBusinessModule.kt
package expo.modules.tiktokbusiness

import android.util.Log
import com.tiktok.TikTokBusinessSdk
import com.tiktok.TikTokBusinessSdk.TTConfig
import com.tiktok.TikTokBusinessSdk.LogLevel
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject

class TiktokBusinessModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TiktokBusiness")

    AsyncFunction("initialize") { config: Map<String, Any> ->
      try {
        val appId = config["appId"] as? String
        val tiktokAppId = config["tiktokAppId"] as? String

        if (appId.isNullOrEmpty() || tiktokAppId.isNullOrEmpty()) {
          return@AsyncFunction false
        }

        val context = appContext.reactContext?.applicationContext
          ?: return@AsyncFunction false

        val debugMode = config["debugMode"] as? Boolean ?: false

        val ttConfig = TTConfig(context)
        ttConfig.setAppId(appId)
        ttConfig.setTTAppId(tiktokAppId)

        if (debugMode) {
          ttConfig.setLogLevel(LogLevel.DEBUG)
        }

        TikTokBusinessSdk.initializeSdk(ttConfig)
        return@AsyncFunction true
      } catch (e: Exception) {
        Log.e("TiktokBusiness", "Error initializing TikTok SDK: ${e.message}")
        return@AsyncFunction false
      }
    }

    AsyncFunction("trackEvent") { eventName: String, eventData: Map<String, Any>? ->
      try {
        val jsonProps = if (eventData != null) JSONObject(eventData) else null
        TikTokBusinessSdk.trackEvent(eventName, jsonProps)
        return@AsyncFunction true
      } catch (e: Exception) {
        Log.e("TiktokBusiness", "Error tracking event: ${e.message}")
        return@AsyncFunction false
      }
    }
  }
}

File: modules/tiktok-business/ios/TiktokBusinessModule.swift
import ExpoModulesCore
import TikTokBusinessSDK

public class TiktokBusinessModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TiktokBusiness")

    AsyncFunction("initialize") { (config: [String: Any], promise: Promise) in
      guard let appId = config["appId"] as? String,
            let tiktokAppId = config["tiktokAppId"] as? String else {
        promise.resolve(false)
        return
      }

      guard let sdkConfig = TikTokConfig(appId: appId, tiktokAppId: tiktokAppId) else {
        promise.resolve(false)
        return
      }

      if let debugMode = config["debugMode"] as? Bool, debugMode {
        sdkConfig.enableDebugMode()
      }

      TikTokBusiness.initializeSdk(sdkConfig) { success, error in
        if !success {
          print("TikTok SDK init failed: \(error?.localizedDescription ?? "unknown")")
        }
        if success && TikTokBusiness.isDebugMode() {
          print("TikTok SDK test event code: \(TikTokBusiness.getTestEventCode() ?? "none")")
        }
        promise.resolve(success)
      }
    }

    Function("getTestEventCode") { () -> String? in
      return TikTokBusiness.getTestEventCode()
    }

    AsyncFunction("trackEvent") { (eventName: String, eventData: [String: Any]) -> Bool in
      TikTokBusiness.trackEvent(eventName, withProperties: eventData)
      return true
    }
  }
}
